bot/exts/utilities/wtf_python.py

```diff
@@ -0,0 +1,154 @@
+import logging
+import random
+import re
+from enum import Enum
+from functools import partial
+from typing import Optional
+
+from discord import Embed
+from discord.ext import commands, tasks
+from rapidfuzz import process
+
+from bot import constants
+from bot.bot import Bot
+
+log = logging.getLogger(__name__)
+
+WTF_PYTHON_RAW_URL = "http://raw.githubusercontent.com/satwikkansal/wtfpython/master/"
+BASE_URL = "https://github.com/satwikkansal/wtfpython"
+FETCH_TRIES = 3
+
+ERROR_MESSAGE = f"""
+Unknown WTF Python Query. Please try to reformulate your query.
+
+**Examples**:
+```md
+{constants.Client.prefix}wtf wild imports
+{constants.Client.prefix}wtf subclass
+{constants.Client.prefix}wtf del
+```
+If the problem persists send a message in <#{constants.Channels.dev_contrib}>
+"""
+
+MINIMUM_CERTAINTY = 50
+
+
+class Action(Enum):
+    """Represents an action to perform on an extension."""
+
+    # Need to be partial otherwise they are considered to be function definitions.
+    LOAD = partial(Bot.load_extension)
+    UNLOAD = partial(Bot.unload_extension)
+
+
+class WTFPython(commands.Cog):
+    """Cog that allows getting WTF Python entries from the WTF Python repository."""
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+        self.fetch_readme.start()
+
+        self.headers: dict[str, str] = dict()
+
+    @tasks.loop(hours=1)
+    async def fetch_readme(self) -> None:
+        """Gets the content of README.md from the WTF Python Repository."""
+        failed_tries = 0
+
+        for x in range(FETCH_TRIES):
+            async with self.bot.http_session.get(f"{WTF_PYTHON_RAW_URL}README.md") as resp:
+                log.trace("Fetching the latest WTF Python README.md")
+
+                if resp.status == 200:
+                    raw = await resp.text()
+                    self.parse_readme(raw)
+                    log.debug(
+                        "Successfully fetched the latest WTF Python README.md, "
+                        "breaking out of retry loop"
+                    )
+                    break
+
+                else:
+                    failed_tries += 1
+                    log.debug(
+                        "Failed to get latest WTF Python README.md on try "
+                        f"{x}/{FETCH_TRIES}. Status code {resp.status}"
+                    )
+
+        if failed_tries == 3:
+            log.error("Couldn't fetch WTF Python README.md after 3 tries, unloading extension.")
+            action = Action.UNLOAD
+        else:
+            action = Action.LOAD
+
+        verb = action.name.lower()
+        ext = "bot.exts.utilities.wtf_python"
+
+        try:
+            action.value(self.bot, ext)
+        except (commands.ExtensionAlreadyLoaded, commands.ExtensionNotLoaded):
+            log.debug(f"Extension `{ext}` is already {verb}ed.")
+        else:
+            log.debug(f"Extension {verb}ed: `{ext}`.")
```

Comment:
Yeah I'm getting stuck on this one. I haven't used `asyncio` yet so maybe that is what is tripping me up. 
Also, how can I get the "last" time the cog was loaded instead of just calling it immediately?

Here is what I have so far:

```py
class WTFPython(commands.Cog):
    """Cog that allows getting WTF Python entries from the WTF Python repository."""

    def __init__(self, bot: Bot):
        self.bot = bot
        self.headers: dict[str, str] = dict()
        self.last_fetched = datetime.datetime.now()  # this is just the current time, how to access last load?
        log.debug(f"{self.last_fetched = }")

        if self.last_fetched + datetime.timedelta(seconds=10) < datetime.datetime.now():  # this will never be true currently
            log.debug("Last loading time was within x amount of time.")
            asyncio.create_task(self.fetch_readme)  # is this correct?

    # cut out most everything from the function, this is what is left
    async def fetch_readme(self) -> None:
        """Gets the content of README.md from the WTF Python Repository."""
        async with self.bot.http_session.get(f"{WTF_PYTHON_RAW_URL}README.md") as resp:
            log.trace("Fetching the latest WTF Python README.md")
            if resp.status == 200:
                raw = await resp.text()
                self.parse_readme(raw)
```

Reply:
I am thinking something more like this:
```python

README_REFRESH = 60  # Amount of minutes

class WTFPython(commands.Cog):
    """Cog that allows getting WTF Python entries from the WTF Python repository."""

    def __init__(self, bot: Bot):
        self.bot = bot
        self.headers: dict[str, str] = dict()
        self.last_fetched = datetime.datetime.now()
        log.debug(f"{self.last_fetched = }")

        asyncio.create_task(self.fetch_readme)  # Start a first task

    async def fetch_readme(self) -> None:
        """Gets the content of README.md from the WTF Python Repository."""
        refresh = self.last_fetched + datetime.timedelta(minutes=README_REFRESH)

        if refresh > datetime.datetime.now() and self.headers:  # We also want to fetch if we haven't previously
            return  # Our cache should be up-to-date

        async with self.bot.http_session.get(f"{WTF_PYTHON_RAW_URL}README.md") as resp:
            log.trace("Fetching the latest WTF Python README.md")
            if resp.status == 200:
                raw = await resp.text()
                self.parse_readme(raw)
```
You then call `asyncio.create_task(self.fetch_readme)` every time someone runs the command.

Reply:
Whoa that is awesome! Thanks for the super clear example.